File: frontend/src/api/observation.ts
import { get } from './client';

// 类型定义

// 异常状态汇总
export interface ObservationSummary {
  podAnomalyCount: number;
  nodeAnomalyCount: number;
  resourceExcessCount: number;
  activeAlertCount: number;
}

// Pod 异常
export interface PodAnomaly {
  name: string;
  namespace: string;
  reason: string;
  message: string;
  restartCount: number;
  duration: string;
  nodeName?: string;
}

// 节点异常
export interface NodeAnomaly {
  name: string;
  reason: string;
  message: string;
  duration: string;
  affectedPods: number;
}

// 资源超限
export interface ResourceExcess {
  type: 'cpu' | 'memory' | 'disk';
  resourceName: string;
  namespace?: string;
  usagePercent: number;
  threshold: number;
}

// 时序数据点
export interface TimeSeriesPoint {
  timestamp: number;
  value: number;
}

// 同比环比结果
export interface ComparisonResult {
  wow: number;    // 周环比
  mom: number;    // 月同比
  trend: 'up' | 'down' | 'stable';
  currentAvg: number;
}

// 资源趋势
export interface ResourceTrend {
  current: TimeSeriesPoint[];
  previous?: TimeSeriesPoint[];
  comparison: ComparisonResult;
}

// 告警趋势数据点
export interface AlertTrendPoint {
  date: string;
  count: number;
}

// 告警趋势
export interface AlertTrend {
  current: AlertTrendPoint[];
  previous?: AlertTrendPoint[];
  comparison: ComparisonResult;
}

// Pod 重启趋势数据点
export interface RestartTrendPoint {
  date: string;
  count: number;
  podName?: string;
}

// Pod 重启趋势
export interface RestartTrend {
  current: RestartTrendPoint[];
  previous?: RestartTrendPoint[];
  comparison: ComparisonResult;
}

// 列表响应
export interface ListResponse<T> {
  items: T[];
  total: number;
}

// 时间范围类型
export type TimeRange = 'realtime' | '1h' | '24h' | '7d' | '30d';

// 资源类型
export type ResourceType = 'cpu' | 'memory';

// API 模块
export const observationApi = {
  // 获取异常状态汇总
  getSummary: () =>
    get<ObservationSummary>('/observation/summary'),

  // 获取异常 Pod 列表
  getPodAnomalies: (params?: { namespace?: string }) =>
    get<ListResponse<PodAnomaly>>('/observation/pods/anomaly', params),

  // 获取异常节点列表
  getNodeAnomalies: () =>
    get<ListResponse<NodeAnomaly>>('/observation/nodes/anomaly'),

  // 获取资源超限列表
  getResourceExcess: (params?: { namespace?: string }) =>
    get<ListResponse<ResourceExcess>>('/observation/resources/excess', params),

  // 获取资源使用趋势
  getResourceTrend: (type: ResourceType, range: TimeRange) =>
    get<ResourceTrend>('/observation/trends/resource', { type, range }),

  // 获取告警趋势
  getAlertTrend: (range?: TimeRange) =>
    get<AlertTrend>('/observation/trends/alerts', range ? { range } : undefined),

  // 获取 Pod 重启趋势
  getRestartTrend: (range?: TimeRange) =>
    get<RestartTrend>('/observation/trends/restarts', range ? { range } : undefined),
};

export default observationApi;

File: backend/internal/api/handlers/observation.go
package handlers

import (
	"context"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/k8s-dashboard/backend/internal/observation"
)

// ObservationHandler 集群观测处理器
type ObservationHandler struct {
	service *observation.Service
}

// NewObservationHandler 创建观测处理器
func NewObservationHandler(service *observation.Service) *ObservationHandler {
	return &ObservationHandler{
		service: service,
	}
}

// GetObservationSummary 获取异常状态汇总
func (h *ObservationHandler) GetObservationSummary(c *gin.Context) {
	ctx := context.Background()

	summary, err := h.service.GetSummary(ctx)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, summary)
}

// GetPodAnomalies 获取异常 Pod 列表
func (h *ObservationHandler) GetPodAnomalies(c *gin.Context) {
	ctx := context.Background()
	namespace := c.Query("namespace")

	anomalies, err := h.service.GetPodAnomalies(ctx, namespace)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"items": anomalies,
		"total": len(anomalies),
	})
}

// GetNodeAnomalies 获取异常节点列表
func (h *ObservationHandler) GetNodeAnomalies(c *gin.Context) {
	ctx := context.Background()

	anomalies, err := h.service.GetNodeAnomalies(ctx)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"items": anomalies,
		"total": len(anomalies),
	})
}

// GetResourceExcess 获取资源超限列表
func (h *ObservationHandler) GetResourceExcess(c *gin.Context) {
	ctx := context.Background()
	namespace := c.Query("namespace")

	excess, err := h.service.GetResourceExcess(ctx, namespace)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"items": excess,
		"total": len(excess),
	})
}

// GetResourceTrend 获取资源使用趋势
func (h *ObservationHandler) GetResourceTrend(c *gin.Context) {
	ctx := context.Background()
	resourceType := observation.ResourceType(c.DefaultQuery("type", "cpu"))
	timeRange := observation.ParseTimeRange(c.DefaultQuery("range", "24h"))

	trend, err := h.service.GetResourceTrend(ctx, resourceType, timeRange)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, trend)
}

// GetAlertTrend 获取告警趋势
func (h *ObservationHandler) GetAlertTrend(c *gin.Context) {
	ctx := context.Background()
	timeRange := observation.ParseTimeRange(c.DefaultQuery("range", "7d"))

	trend, err := h.service.GetAlertTrend(ctx, timeRange)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, trend)
}

// GetRestartTrend 获取 Pod 重启趋势
func (h *ObservationHandler) GetRestartTrend(c *gin.Context) {
	ctx := context.Background()
	timeRange := observation.ParseTimeRange(c.DefaultQuery("range", "24h"))

	trend, err := h.service.GetRestartTrend(ctx, timeRange)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, trend)
}
